platforms/jvm/plugins-application/src/main/java/org/gradle/api/internal/plugins/StartScriptTemplateBindingFactory.java

```diff
@@ -63,24 +72,59 @@ public Map<String, String> transform(JavaAppStartScriptGenerationDetails details
 
     }
 
+    private static String getMainClassName(AppEntryPoint entryPoint, String entryPointArgs) {
+        if (entryPoint instanceof MainClass) {
+            return ((MainClass) entryPoint).getMainClassName();
+        } else if (entryPoint instanceof MainModule) {
+            // For legacy reasons, keep the mainClassName as the module invocation for scripts which used it that way
+            return entryPointArgs;
+        } else {
+            return "";
+        }
+    }
+
+    private static AppEntryPoint getEntryPoint(JavaAppStartScriptGenerationDetails details) {
+        if (details instanceof DefaultJavaAppStartScriptGenerationDetails) {
+            return ((DefaultJavaAppStartScriptGenerationDetails) details).getEntryPoint();
+        } else {
+            // Provide compatibility in case someone was manually implementing JavaAppStartScriptGenerationDetails
+            return new MainClass(details.getMainClassName());
+        }
+    }
+
+    private String encodeEntryPoint(AppEntryPoint entryPoint) {
+        if (entryPoint instanceof MainClass) {
+            return ((MainClass) entryPoint).getMainClassName();
+        } else if (entryPoint instanceof MainModule) {
+            return "--module " + getModuleEntryPoint((MainModule) entryPoint);
+        } else if (entryPoint instanceof ExecutableJar) {
+            // We need to also escape quotes in the JAR path, so our quotes aren't broken by the JAR path
+            // In theory we should be doing this in `encodePath`, but I wanted to avoid making behavioral changes to `classpath` and `modulePath`
+            String jarPathEscaped = encodePath(((ExecutableJar) entryPoint).getJarPath()).replace("\"", "\\\"");
+            return "-jar \"" + jarPathEscaped + "\"";
```

Comment:
Should this just be a part of `AppEntryPoint`?

e.g., `AppEntryPoint.toArgs()`